src/projectM-emscripten/projectM_SDL_emscripten.cpp

```diff
@@ -5,224 +5,275 @@
  */
 
 #include <math.h>
-#include <projectM.hpp>
+#include <projectM.h>
+#include <sdltoprojectM.h>
 
-#ifdef __APPLE__
-#include <OpenGL/gl.h>
-#include "SDL.h"
-#elif EMSCRIPTEN
 #include <emscripten.h>
 #include <GL/gl.h>
 #include <SDL.h>
-#endif
+
+#include <string>
+
+#include <dirent.h>
 
 const float FPS = 60;
 
-typedef struct {
-    projectM *pm;
-    SDL_Window *win;
-    SDL_Renderer *rend;
-    bool done;
-    projectM::Settings settings;
-    SDL_AudioDeviceID audioInputDevice;
-} projectMApp;
+struct projectMApp
+{
+    projectm_handle pm{ nullptr };
+    SDL_Window* win{ nullptr };
+    SDL_Renderer* rend{ nullptr };
+    bool done{ false };
+    projectm_settings settings;
+    SDL_AudioDeviceID audioInputDevice{ 0 };
+    int audioInputDevicesCount{ 0 };
+    int audioChannelsCount{ 0 };
+    bool isFullscreen{ false };
+};
 
 projectMApp app;
 
-int selectAudioInput(projectMApp *app) {
-    int i, count = SDL_GetNumAudioDevices(0);  // param=isCapture (not yet functional)
+void audioInputCallbackF32(void* userdata, unsigned char* stream, int len)
+{
+    if (app.audioChannelsCount == 1)
+    {
+        projectm_pcm_add_float_1ch_data(app.pm, reinterpret_cast<float*>(stream), len / sizeof(float));
+    }
+    else if (app.audioChannelsCount == 2)
+    {
+        projectm_pcm_add_float_2ch_data(app.pm, reinterpret_cast<float*>(stream), len / sizeof(float));
+    }
+    else
+    {
+        SDL_LogError(SDL_LOG_CATEGORY_APPLICATION, "Multichannel audio not supported");
+    }
+}
+
+bool selectAudioInput(projectMApp* application)
+{
+    app.audioInputDevicesCount = SDL_GetNumAudioDevices(1);
 
-    if (! count) {
-        fprintf(stderr, "No audio input capture devices detected\n");
-        return 0;
+    if (!app.audioInputDevicesCount)
+    {
+        fprintf(stderr, "No audio input capture devices detected. Faking audio using random data.\n");
+        return false;
     }
 
-    printf("count: %d\n", count);
-    for (i = 0; i < count; ++i) {
-        printf("Audio device %d: %s\n", i, SDL_GetAudioDeviceName(i, 0));
+    printf("count: %d\n", app.audioInputDevicesCount);
```

Comment:
```suggestion
```

Reply:
Count should always be 1 in browser environments, so yeah, this output is rather unnecessary. No matter how many mics are connected, I always got a single "System Audio Device" in the enumeration, even if the user hasn't allowed mic access. If no access is allowed, the audio data provided by SDL is just zeroes.